File: AOM/packages/aom-protocol-ts/src/raw.ts
export type JsonValue =
  | null
  | boolean
  | number
  | string
  | JsonValue[]
  | { [key: string]: JsonValue };

export type RawEventType =
  | "surface_snapshot"
  | "surface_click"
  | "surface_text_input"
  | "lifecycle"
  | "navigation"
  | "network_request"
  | "network_response"
  | "storage_read"
  | "storage_write"
  | "state_change"
  | "error";

export interface RawEventSource {
  adapterId: string;
  probeId: string;
  sourceType: string;
}

export interface RawRef {
  rawId: string;
  kind?: string;
  label?: string;
}

export interface RawArtifactDescriptor {
  artifactId: string;
  kind: string;
  locator: string;
  format?: string;
  digest?: string;
  metadata: Record<string, JsonValue>;
}

export interface RawStaticNode {
  rawId: string;
  kind: string;
  label?: string;
  artifactId: string;
  artifactOffset?: string;
  attributes: Record<string, JsonValue>;
  evidenceIds: string[];
}

export interface RawStaticEdge {
  rawId: string;
  fromRawId: string;
  toRawId: string;
  relationship: string;
  evidenceIds: string[];
}

export interface RawStaticSnapshot {
  snapshotId: string;
  targetId: string;
  platform: string;
  timestamp: number;
  adapterId: string;
  artifacts: RawArtifactDescriptor[];
  nodes: RawStaticNode[];
  edges: RawStaticEdge[];
  evidenceIds: string[];
}

export interface RawEvent {
  eventId: string;
  targetId: string;
  platform: string;
  timestamp: number;
  sequence: number;
  type: RawEventType;
  source: RawEventSource;
  subject?: RawRef;
  object?: RawRef;
  payload: Record<string, JsonValue>;
  evidenceIds: string[];
}

export interface RawRuntimeNode {
  rawId: string;
  kind: string;
  role?: string;
  label?: string;
  value?: JsonValue;
  attributes: Record<string, JsonValue>;
  children: string[];
}

export interface RawRuntimeSnapshot {
  snapshotId: string;
  targetId: string;
  platform: string;
  timestamp: number;
  nodes: RawRuntimeNode[];
  evidenceIds: string[];
}

export type RawActionType = "click" | "set_text" | "scroll" | "back" | "wait_for";

export interface RawAction {
  actionId: string;
  targetId: string;
  type: RawActionType;
  targetRawId?: string;
  params: Record<string, JsonValue>;
}

export interface RawActionResult {
  actionId: string;
  targetId: string;
  ok: boolean;
  errorCode?: string;
  message?: string;
  evidenceIds: string[];
}

File: AOM/crates/aom-protocol-rs/src/raw.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::BTreeMap;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum RawEventType {
    SurfaceSnapshot,
    SurfaceClick,
    SurfaceTextInput,
    Lifecycle,
    Navigation,
    NetworkRequest,
    NetworkResponse,
    StorageRead,
    StorageWrite,
    StateChange,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RawEventSource {
    pub adapter_id: String,
    pub probe_id: String,
    pub source_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RawRef {
    pub raw_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub kind: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RawArtifactDescriptor {
    pub artifact_id: String,
    pub kind: String,
    pub locator: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub format: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub digest: Option<String>,
    #[serde(default)]
    pub metadata: BTreeMap<String, Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RawStaticNode {
    pub raw_id: String,
    pub kind: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
    pub artifact_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub artifact_offset: Option<String>,
    #[serde(default)]
    pub attributes: BTreeMap<String, Value>,
    #[serde(default)]
    pub evidence_ids: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RawStaticEdge {
    pub raw_id: String,
    pub from_raw_id: String,
    pub to_raw_id: String,
    pub relationship: String,
    #[serde(default)]
    pub evidence_ids: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RawStaticSnapshot {
    pub snapshot_id: String,
    pub target_id: String,
    pub platform: String,
    pub timestamp: u64,
    pub adapter_id: String,
    #[serde(default)]
    pub artifacts: Vec<RawArtifactDescriptor>,
    #[serde(default)]
    pub nodes: Vec<RawStaticNode>,
    #[serde(default)]
    pub edges: Vec<RawStaticEdge>,
    #[serde(default)]
    pub evidence_ids: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RawEvent {
    pub event_id: String,
    pub target_id: String,
    pub platform: String,
    pub timestamp: u64,
    pub sequence: u64,
    #[serde(rename = "type")]
    pub event_type: RawEventType,
    pub source: RawEventSource,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subject: Option<RawRef>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub object: Option<RawRef>,
    #[serde(default)]
    pub payload: BTreeMap<String, Value>,
    #[serde(default)]
    pub evidence_ids: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RawRuntimeNode {
    pub raw_id: String,
    pub kind: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub role: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value: Option<Value>,
    #[serde(default)]
    pub attributes: BTreeMap<String, Value>,
    #[serde(default)]
    pub children: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RawRuntimeSnapshot {
    pub snapshot_id: String,
    pub target_id: String,
    pub platform: String,
    pub timestamp: u64,
    #[serde(default)]
    pub nodes: Vec<RawRuntimeNode>,
    #[serde(default)]
    pub evidence_ids: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum RawActionType {
    Click,
    SetText,
    Scroll,
    Back,
    WaitFor,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RawAction {
    pub action_id: String,
    pub target_id: String,
    #[serde(rename = "type")]
    pub action_type: RawActionType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_raw_id: Option<String>,
    #[serde(default)]
    pub params: BTreeMap<String, Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RawActionResult {
    pub action_id: String,
    pub target_id: String,
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
    #[serde(default)]
    pub evidence_ids: Vec<String>,
}
